pkg/planner/core/memtable_infoschema_extractor.go

```diff
@@ -0,0 +1,396 @@
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+) ([]*model.TableInfo, error) {
+	tableNames := e.getSchemaObjectNames(_TableName)
+	partIDs := e.getSchemaObjectNames(_TIDBPartitionID)
+	if len(tableNames)+len(partIDs) == 0 {
+		return is.SchemaTableInfos(ctx, schema)
+	}
+
+	tables := make(map[int64]*model.TableInfo, 8)
+	err := findNameAndAppendToTableMap(ctx, is, schema, tableNames, tables)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	findByPartIDAndAppendToTableMap(is, schema, partIDs, tables)
+	return maps.Values(tables), nil
+}
+
+// InfoSchemaStatisticsExtractor is the predicate extractor for  information_schema.statistics.
+type InfoSchemaStatisticsExtractor struct {
+	InfoSchemaBaseExtractor
+}
+
+// InfoSchemaSchemataExtractor is the predicate extractor for information_schema.schemata.
+type InfoSchemaSchemataExtractor struct {
+	InfoSchemaBaseExtractor
+}
+
+// ListSchemas lists related schemas from predicate.
+// If no schema found in predicate, it return all schemas.
+func (e *InfoSchemaSchemataExtractor) ListSchemas(is infoschema.InfoSchema) []model.CIStr {
+	return e.listSchemas(is, _SchemaName)
+}
+
+func (e *InfoSchemaBaseExtractor) listSchemas(is infoschema.InfoSchema, schemaCol string) []model.CIStr {
+	schemas := e.getSchemaObjectNames(schemaCol)
+	if len(schemas) == 0 {
+		ret := is.AllSchemaNames()
+		slices.SortFunc(ret, func(a, b model.CIStr) int {
+			return strings.Compare(a.L, b.L)
+		})
+		return ret
+	}
+	ret := schemas[:0]
+	for _, s := range schemas {
+		if n, ok := is.SchemaByName(s); ok {
+			ret = append(ret, n.Name)
+		}
+	}
+	return ret
+}
+
+func findNameAndAppendToTableMap(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+	tableNames []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) error {
+	for _, n := range tableNames {
+		tbl, err := is.TableByName(ctx, schema, n)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
+			}
+			return errors.Trace(err)
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+	return nil
+}
+
+func findIDAndAppendToTableMap(
+	ctx context.Context,
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+	tableIDs []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) error {
+	for _, id := range parseIDs(tableIDs) {
+		tbl, ok := is.TableByID(id)
+		if !ok {
+			continue
+		}
+		_, err := is.TableByName(ctx, schema, tbl.Meta().Name)
+		if err != nil {
+			if terror.ErrorEqual(err, infoschema.ErrTableNotExists) {
+				continue
+			}
+			return errors.Trace(err)
+		}
+		tblInfo := tbl.Meta()
+		tables[tblInfo.ID] = tblInfo
+	}
+	return nil
+}
+
+func findByPartIDAndAppendToTableMap(
+	is infoschema.InfoSchema,
+	schema model.CIStr,
+	partIDs []model.CIStr,
+	tables map[int64]*model.TableInfo,
+) {
+	for _, pid := range parseIDs(partIDs) {
+		tbl, db, _ := is.FindTableByPartitionID(pid)
+		if tbl == nil {
+			continue
+		}
+		if db.Name.L != schema.L {
```

Comment:
if there's no selection filter on database in SQL, in current implementation we still iterate every dbdatabaseand double the partition's database? I think we can query partition first and then filter the database once.

Reply:
seems this problem also exists for table_id selection (I guess infoschema know the database of a table when query by ID?)

Reply:
maybe run ID-based selection first. I guess optimizer will choose UNIQUE INDEX prior to other INDEX because it will return only one row at most.

Reply:
https://github.com/pingcap/tidb/pull/55144/commits/675c591605d451213878c5f69d25e6f268e82278